### PR TITLE
[nfc][docs]Generalize header description and ascii art for indexed profiles

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
- Add pointers to code for source of truth.
- Move necessary details from doc to code.

### CODE REVIEW

**Comment by minglotus-6** [comment]
I think the doc is easier to maintain by removing boilerplates. Otherwise it's painful to update ascii art graph with templated content.

Planned to submit this change before https://github.com/llvm/llvm-project/pull/83309.

File: llvm/docs/InstrProfileFormat.rst [hunk 1]
```diff
@@ -369,55 +373,45 @@ General Storage Layout
...
-            |   |   |    |  +-----------------------+---+
-            |   |   |    |  |   Profile Summary     |   |
-            |   |   |    |  +-----------------------+   P
-            |   |   +------>|    Function data      |   A
-            |   |        |  +-----------------------+   Y
-            |   +---------->|  MemProf profile data |   L
-            |            |  +-----------------------+   O
-            |            +->|    Binary Ids         |   A
+                            |       Byte Offset     |   A
+                    +------ |      of section 2     |   D
```
Comment (david-xl) [inline llvm/docs/InstrProfileFormat.rst:377]: why starting the number from 2?

File: llvm/docs/InstrProfileFormat.rst [hunk 2]
```diff
@@ -369,55 +373,45 @@ General Storage Layout
...
-
-``Version``
-  Similar to raw profile version, the lower 32 bits specify the version of the
-  indexed profile and the most significant 32 bits are reserved to specify the
-  variant types of the profile.
-
-``HashType``
-  The hashing scheme for on-disk hash table keys. Only MD5 hashing is used as of
-  writing.
+The `Header struct`_ should be the source of truth. Field names and comments
```
Comment (david-xl) [inline llvm/docs/InstrProfileFormat.rst:402]: Probably document the sections are A, B, C, and D, and their order is defined in Header struct.

Comment (minglotus-6) [reply llvm/docs/InstrProfileFormat.rst]: Incorporate the section names in the graph above and reworded the paragraph slightly.

For indexed profiles, the byte offset is relative to the start of profile memory buffer. Thereby the field order in the struct definition doesn't need to be the same as the order of payloads in the profile (e.g., if we need it sometime in the future).

**Comment by minglotus-6** [comment]
Bump.

**Review by david-xl — APPROVED** [review]

**Pull request merged**